### PR TITLE
Render checkbox question with Thymeleaf

### PR DESCRIPTION
**Pull request by greg-hyde**
### Description

Implement rendering of checkbox question using Thymeleaf and USWDS components

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers).

#### User visible changes

- [x] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [x] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [x] Manually tested at 200% size
- [x] Manually evaluated tab order

#### New Features

- [x] Conditioned new functionality on a [FeatureFlag](https://github.com/civiform/civiform/wiki/Feature-Flags)
- [x] Wrote browser tests with the feature flag off and on, etc.

### CODE REVIEW

**Comment by greg-hyde** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/civiform/civiform/6702?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#6703** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6703?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6702** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6702?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6701** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6701?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6700** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6700?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6699** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6699?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6513** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6513?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 3 other dependent PRs ([#6514](https://github.com/civiform/civiform/pull/6514) <a href="https://app.graphite.dev/github/pr/civiform/civiform/6514?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#6531](https://github.com/civiform/civiform/pull/6531) <a href="https://app.graphite.dev/github/pr/civiform/civiform/6531?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#6581](https://github.com/civiform/civiform/pull/6581) <a href="https://app.graphite.dev/github/pr/civiform/civiform/6581?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `main`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/civiform/civiform/6702?utm_source=stack-comment).

File: server/app/views/applicant/ApplicantQuestionFragment.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<div th:fragment="applicant-question (question, messages)">
+    <th:block th:switch="${question.getType().getLabel()}">
```
Comment (greg-hyde) [inline server/app/views/applicant/ApplicantQuestionFragment.html:2]: Annoying thing here is that I've not been able to access static values (like an Enum value) in Thymeleaf because it can't find the relevant class, so for now I put in this string comparison that makes me sad :(

File: server/app/views/questiontypes/QuestionBaseFragment.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<th:block th:fragment="validationErrors (validationErrors, messages)">
+    <div th:each="error: ${validationErrors}" th:text="${error.getMessage(messages)}"></div>
```
Comment (avaleske) [inline server/app/views/questiontypes/QuestionBaseFragment.html:2]: does something like this work?
```
<div th:each="error: ${validationErrors}" th:text="#{${error.key}(${error.args})}"></div>
```
From the last example under https://www.thymeleaf.org/doc/tutorials/3.1/usingthymeleaf.html#messages

File: server/app/views/questiontypes/CheckboxQuestionFragment.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<div th:fragment="checkbox (question, messages)" th:with="questionId=${#strings.randomAlphanumeric(8)}" th:id="${questionId}">
+    <fieldset class="usa-fieldset" th:aria-describedby="${questionId} + '-description'" th:with="multiSelectQuestion=${question.createMultiSelectQuestion()}">
```
Comment (avaleske) [inline server/app/views/questiontypes/CheckboxQuestionFragment.html:2]: Do you have opinions about using the `*{...}` syntax for anywhere below here that references `multiSelectQuestion`?

It seems nice to avoid repetition, but also like it could be hard to follow

File: server/app/views/questiontypes/CheckboxQuestionFragment.html [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<div th:fragment="checkbox (question, messages)" th:with="questionId=${#strings.randomAlphanumeric(8)}" th:id="${questionId}">
```
[the remark this answers is not in the document]
Comment (greg-hyde) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:1]: the random string is used for the ID of the html element, not to be confused with the underlying question. It's mostly used so we can do stuff like aria-describedby or for or other indications for screen readers associating multiple elements.

File: server/app/views/applicant/ApplicantQuestionFragment.html (hunk 1 above)
Comment (greg-hyde) [reply server/app/views/applicant/ApplicantQuestionFragment.html:2]: yes I tried that (which doesn't work because we don't use spring boot). Also tried the @ syntax referenced here https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#preprocessing but no luck. Not giving up yet, since it seems like there has to be a way to do it, but haven't managed it yet

File: server/app/views/questiontypes/CheckboxQuestionFragment.html (hunk 3 above)
Comment (greg-hyde) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:2]: I can go either way on that. I have no objections to the * syntax, especially when reusing something. The only thing that's confusing here is that we're using both question and checkboxQuestion, but it's honestly a bit confusing either way...

File: server/app/views/questiontypes/QuestionBaseFragment.html (hunk 2 above)
Comment (greg-hyde) [reply server/app/views/questiontypes/QuestionBaseFragment.html:2]: good idea I got a variant of this working

File: server/app/views/applicant/ApplicantQuestionFragment.html (hunk 1 above)
Comment (dkatzz) [reply server/app/views/applicant/ApplicantQuestionFragment.html:2]: Also played around with this a bit with no luck :( I tried adding the question type enum class to the template and it showed if I print it out in `<pre th:text="${#vars}"></pre>`, but couldn't figure out how to use it. Super annoying

File: server/app/views/applicant/NorthStarApplicantBaseView.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  protected ThymeleafModule.PlayThymeleafContext createThymeleafContext(Request request) {
+    ThymeleafModule.PlayThymeleafContext context = playThymeleafContextFactory.create(request);
+    context.setVariable("tailwindStylesheet", assetsFinder.path("stylesheets/tailwind.css"));
+    context.setVariable("uswdsStylesheet", assetsFinder.path("dist/uswds.min.css"));
+    context.setVariable("adminJsBundle", assetsFinder.path("dist/admin.bundle.js"));
+    context.setVariable("uswdsJsInit", assetsFinder.path("javascripts/uswds/uswds-init.min.js"));
+    context.setVariable("uswdsJsBundle", assetsFinder.path("dist/uswds.bundle.js"));
+    context.setVariable("ApiDocsController", controllers.api.routes.ApiDocsController);
```
Comment (dkatzz) [inline server/app/views/applicant/NorthStarApplicantBaseView.java:35]: **question:** What is this used for?

File: server/app/views/questiontypes/CheckboxQuestionFragment.html (hunk 3 above)
Comment (greg-hyde) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:2]: Alright I played with this some more and I find that *{} is limiting in instances like this. You can't pass any arguments when using * syntax, so you can't really convert:
${multiSelectQuestion.getValidationErrors().get(question.getContextualizedPath())} OR 
${multiSelectQuestion.optionIsSelected(option)}

so for those you have to do:
${#object.optionIsSelected(option)} instead, and then you're mixing *s and #objects and it all ends up more muddled. It's probably good in other cases though.

Comment (avaleske) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:2]: Mmmm yeah that's annoying if it's mixed.

I can imagine refactors to make it work nicer, like having the selection status on the option itself so we don't have to ask the question if it's the selected option, but then that has other downsides.

File: server/app/views/applicant/NorthStarApplicantBaseView.java (hunk 5 above)
Comment (greg-hyde) [reply server/app/views/applicant/NorthStarApplicantBaseView.java]: it's not used I removed it

File: server/app/views/questiontypes/CheckboxQuestionFragment.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<div th:fragment="checkbox (question)" th:with="questionId=${#strings.randomAlphanumeric(8)}" th:id="${questionId}">
+    <fieldset class="usa-fieldset" th:aria-describedby="${questionId} + '-description'" th:with="multiSelectQuestion=${question.createMultiSelectQuestion()}">
+        <!-- Question title -->
```
[the remark this answers is not in the document]
Comment (greg-hyde) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html]: yeah good point. right now there is some mild variation (legend vs label) but they don't have to be

File: server/app/views/questiontypes/CheckboxQuestionFragment.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
+<div th:fragment="checkbox (question)" th:with="questionId=${#strings.randomAlphanumeric(8)}" th:id="${questionId}">
+    <fieldset class="usa-fieldset" th:aria-describedby="${questionId} + '-description'" th:with="multiSelectQuestion=${question.createMultiSelectQuestion()}">
+        <!-- Title and Help Text -->
+        <div th:replace="~{questiontypes/QuestionBaseFragment :: titleAndHelpText(${question}, ${questionId})}"></div>
```
Comment (dkatzz) [inline server/app/views/questiontypes/CheckboxQuestionFragment.html:4]: **thought:** I don't know if this makes sense based on it being within the div / fieldset, but we could also consider moving this into the ApplicantQuestionFragment file if it will be present in every question?

File: server/app/views/applicant/ApplicantProgramBlockEditTemplate.html [hunk 8]
```diff
@@ -1,13 +1,20 @@
 <!DOCTYPE html>
 <html>
     <head th:replace="~{applicant/ApplicantBaseFragment :: pageHeader}"></head>
-    <body>
+    <!-- TODO: This data attribute is waited for in screenshot tests. Update once we have modal dialogs -->
```
Comment (dkatzz) [inline server/app/views/applicant/ApplicantProgramBlockEditTemplate.html:4]: **suggestion:** One of these todos have your username - consider making these consistent to make it easier to find later?

**Review by dkatzz — APPROVED** [review]
So exciting!

File: server/app/views/applicant/ApplicantProgramBlockEditTemplate.html (hunk 8 above)
Comment (gwendolyngoetz) [reply server/app/views/applicant/ApplicantProgramBlockEditTemplate.html:4]: Austin once reminded me that if the TODO is going to live for a while it's a good idea to create an issue and include the issue number

Comment (dkatzz) [reply server/app/views/applicant/ApplicantProgramBlockEditTemplate.html:4]: Yeah that's even better :)

File: server/app/views/questiontypes/CheckboxQuestionFragment.html (hunk 7 above)
Comment (greg-hyde) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:4]: Se we can't move the validation errors since that varies per question type. I think we could probably move the outer div and the title/help text, but I'd prefer to try that refactor in another PR (it will be a bit easier to visualize once we have a couple question types in)

File: server/app/views/applicant/ApplicantQuestionFragment.html [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<div th:fragment="applicant-question (question)">
+    <th:block th:switch="${question.getType().getLabel()}">
```
[the remark this answers is not in the document]
Comment (greg-hyde) [reply server/app/views/applicant/ApplicantQuestionFragment.html:2]: hmm it's an interesting idea. On the one hand i hate comparing the label string. On the other, it feels a little weird to add something that's 100% view (what file is the thymeleaf file for this fragment and what is the fragment name) into a model class. I think I prefer not to pollute the model class, but I added a TODO for further investigation into accessing enum values

File: server/app/views/questiontypes/CheckboxQuestionFragment.html (hunk 7 above)
Comment (dkatzz) [reply server/app/views/questiontypes/CheckboxQuestionFragment.html:4]: Sounds good, thanks!

File: server/app/views/applicant/ApplicantQuestionFragment.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (dkatzz) [reply server/app/views/applicant/ApplicantQuestionFragment.html:2]: Yeah that's true - plan SGTM!

File: server/app/views/applicant/ApplicantProgramBlockEditTemplate.html (hunk 8 above)
Comment (greg-hyde) [reply server/app/views/applicant/ApplicantProgramBlockEditTemplate.html:4]: done. This is the motivation I needed to finally make an epic and start making issues in github for tasks :)
